x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineAssertions.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ccr.index.engine;
+
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.index.VersionType;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.seqno.SequenceNumbers;
+import org.elasticsearch.rest.RestStatus;
+
+/**
+ * Moved preflight and assertPrimaryIncomingSequenceNumber check to its own class,
+ * so that when testing writing directly into a follower index,
+ * only these assertions here need to be disabled instead of all assertions in FollowingEngine class.
+ */
+final class FollowingEngineAssertions {
+
+    static void preFlight(final Engine.Operation operation) {
+        /*
+         * We assert here so that this goes uncaught in unit tests and fails nodes in standalone tests (we want a harsh failure so that we
+         * do not have a situation where a shard fails and is recovered elsewhere and a test subsequently passes). We throw an exception so
+         * that we also prevent issues in production code.
+         */
+        assert operation.seqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO;
+        if (operation.seqNo() == SequenceNumbers.UNASSIGNED_SEQ_NO) {
+            throw new ElasticsearchStatusException("a following engine does not accept operations without an assigned sequence number",
+                RestStatus.FORBIDDEN);
+        }
+        assert (operation.origin() == Engine.Operation.Origin.PRIMARY) == (operation.versionType() == VersionType.EXTERNAL) :
+            "invalid version_type in a following engine; version_type=" + operation.versionType() + "origin=" + operation.origin();
+    }
+
+    static void assertPrimaryIncomingSequenceNumber(final Engine.Operation.Origin origin, final long seqNo) {
```

Comment:
Can we make this method return true and call `assert assertPrimaryIncomingSequenceNumber` in the FollowingEngine.